packages/protocol/contracts/stability/Exchange.sol

```diff
@@ -107,17 +107,95 @@ contract Exchange is
    * @return The amount of buyToken that was transfered
    * @dev This function can be frozen using the Freezable interface.
    */
+  function sell(uint256 sellAmount, uint256 minBuyAmount, bool sellGold)
+    external
+    onlyWhenNotFrozen
+    updateBucketsIfNecessary
+    nonReentrant
+    returns (uint256)
+  {
+    return _sell(sellAmount, minBuyAmount, sellGold);
+  }
+
+  /**
+   * @dev Deprecated. Exchanges sellAmount of sellToken in exchange for at
+   * least minBuyAmount of buyToken.
+   * Requires the sellAmount to have been approved to the exchange
+   * @param sellAmount The amount of sellToken the user is selling to the exchange
+   * @param minBuyAmount The minimum amount of buyToken the user has to receive for this
+   * transaction to succeed
+   * @param sellGold `true` if gold is the sell token
+   * @return The amount of buyToken that was transfered
+   * @dev This function can be frozen using the Freezable interface.
+   */
   function exchange(uint256 sellAmount, uint256 minBuyAmount, bool sellGold)
     external
     onlyWhenNotFrozen
     updateBucketsIfNecessary
     nonReentrant
     returns (uint256)
+  {
+    return _sell(sellAmount, minBuyAmount, sellGold);
+  }
+
+  /**
+   * @dev Exchanges sellAmount of sellToken in exchange for at least minBuyAmount of buyToken
+   * Requires the sellAmount to have been approved to the exchange
+   * @param sellAmount The amount of sellToken the user is selling to the exchange
+   * @param minBuyAmount The minimum amount of buyToken the user has to receive for this
+   * transaction to succeed
+   * @param sellGold `true` if gold is the sell token
+   * @return The amount of buyToken that was transfered
+   */
+  function _sell(uint256 sellAmount, uint256 minBuyAmount, bool sellGold)
+    private
+    returns (uint256)
   {
     uint256 buyAmount = _getBuyTokenAmount(sellAmount, sellGold);
 
     require(buyAmount >= minBuyAmount, "Calculated buyAmount was less than specified minBuyAmount");
 
+    _exchangeAmounts(sellAmount, buyAmount, sellGold);
+    return buyAmount;
+  }
+
+  /**
+   * @dev Exchanges buyAmount of buyToken in exchange for at most maxSellAmount of sellToken
+   * Requires the sellAmount to have been approved to the exchange
+   * @param buyAmount The amount of buyToken the user is buying from the exchange
+   * @param maxSellAmount The maximum amount of sellToken the user can sell for this
+   * transaction to succeed
+   * @param buyGold `true` if gold is the buy token
+   * @return The amount of sellToken that was sold
+   * @dev This function can be frozen using the Freezable interface.
+   */
+  function buy(uint256 buyAmount, uint256 maxSellAmount, bool buyGold)
```

Comment:
The way `buy` and `sell` work solves everything discussed in https://github.com/celo-org/celo-proposals/pull/26. To increase usability even further and mimic how the trading interface a centralized exchanges usually looks like, we could think about additionally offering functions

`buyCELO(uint256 buyAmount, uint256 maxPrice)`

and 

`sellCELO(uint256 sellAmount, uint256 minPrice)`

where the price parameter is always the number of cUSD (paid/received) per CELO. Allowing folks to specify the price directly avoids that they need to calculate the minBuyAmount or maxSellAmount themselves from the price they target. WDYT?

Reply:
Thought about this again - I think it's totally fine to just stay with `buy` and `sell` for now and add the above functions later if we at some point feel like we need to improve usability further.

Reply:
The above interface can (and should) be implemented at the contractKit level to save gas